### ISSUE TITLE
Bug:  Component with a Symbol as key, causes Crash

### ISSUE THREAD

**Issue by omarsy** [facebook/react#19851]
React version:
16.13.1
# Steps To Reproduce

1. Go To the link https://codesandbox.io/s/happy-ramanujan-xlegp?file=/src/App.js
2. We can see errors caused by this part of code: https://github.com/facebook/react/blob/6fddca27e75950adda92ab4f4946442907dc3bb7/packages/react/src/ReactElement.js#L228

Link to code example:
https://codesandbox.io/s/happy-ramanujan-xlegp?file=/src/App.js

## The current behavior
Crash
## The expected behavior
No crash

**Comment by bpernick** [comment]
So I guess the 3 options I see would be:
1. Disallow Symbol keys but handle them with a graceful error message that doesn't crash the app
2. Allow Symbol keys but convert them to a string with Symbol.toString()
3. (possibly) Allow Symbol keys to remain as symbols

I think I'd lean toward option 2 -- what do y'all think?

**Comment by kunukn** [comment]
A possible solution for option 2?

```js
if (hasValidKey(config)) {
  key = typeof config.key === 'symbol' 
    ? config.key.toString()
    : '' + config.key;
}
```

**Comment by bpernick** [comment]
@kunukn you read my mind. Working on that pr now :)

**Comment by bvaughn** [comment]
FWIW, `key`s *should be strings*. The `'' + key` fallback is a convenience concession to indices-as-keys

**Comment by bpernick** [comment]
Hmm thanks Brian! It looks like fully supporting symbol keys is unlikely to be useful and likely to be dangerous because it could circumvent React's checks for duplicate keys among siblings.

Rather than actually supporting Symbol keys I just tried to make sure they were automatically stringified. The 3 advantages I see are:

1. Symbol keys are treated the same way as number keys, obj keys, etc.
2. Symbol keys don't automatically crash the app.
3. If the dev does something bad, like " return (<><div key={ Symbol('foo') } ><div key={ Symbol('foo') } ></> ", they will get a descriptive error about duplicate keys, rather than the arcane one Omar received.

**Comment by bpernick** [comment]
Let me know what you think. If we decide no, perhaps another solution would be to disallow symbols explicitly in hasValidKey?

**Comment by bvaughn** [comment]
My initial inclination would be to add a DEV-only check + warning that symbols-as-keys isn't supported.

That being said, I don't really speak for the team here. :smile:

**Comment by bpernick** [comment]
@bvaughn Ok! Well I linked what I got for reference and happy to change it if that's what we decide

**Comment by juanjorge** [comment]
> Hmm thanks Brian! It looks like fully supporting symbol keys is unlikely to be useful and likely to be dangerous because it could circumvent React's checks for duplicate keys among siblings.
> 
> Rather than actually supporting Symbol keys I just tried to make sure they were automatically stringified. The 3 advantages I see are:
> 
> 1. Now Symbol keys are treated the same way as number keys, etc.
> 2. Symbol keys don't automatically crash the app.
> 3. If the dev does something bad, like " return (<><div key={ Symbol('foo') } ><div key={ Symbol('foo') } ></> ", they will get a descriptive error about duplicate keys, rather than the arcane one Omar received.
> 
> Let me know what you think -- If we decide no, perhaps another solution would be to disallow symbols explicitly in hasValidKey

The third point is not true, you won't receive an error, on that case you will receive 2 different values.

**Comment by bvaughn** [comment]
>> If the dev does something bad, like " return (<><div key={ Symbol('foo') } ><div key={ Symbol('foo') } ></> ", they will get a descriptive error about duplicate keys, rather than the arcane one Omar received.
>
> The third point is not true, you won't receive an error, on that case you will receive 2 different values.

The context of the above comment is that if React called `toString` on symbols, then two symbols which were not equal before (`Symbol('foo') !== (`Symbol('foo')`) would now be equal (`"Symbol(foo)" === "Symbol(foo)"`)

**Comment by bvaughn** [comment]
We would also need to consider how Symbols work with server side rendering (how they get serialized and sent to the client). For example, even if React handled the serialization– how would it know how to reconstitute a symbol on the client? Should it use `Symbol("...")` or `Symbol.for("...")`? Each one is significantly different in terms of behavior.

**Comment by juanjorge** [comment]
yeah, that's a good point, the `hydrate`, in the global context, should get it using `Symbol.for("...")` in the client, but should be tested, im not sure..

**Comment by bvaughn** [comment]
I agree that using `Symbol.for("...")` would probably be the better option, but there's the possibility that `Symbol("...")` was used intentionally (to create unique symbols regardless of their string content) at which point– using `Symbol.for("...")` during hydration would almost certainly cause errors.

I don't know if this was part of the reason why symbols have never been supported as keys 😄  I don't think I've ever had a conversation about this with other team members. Does seem like an added consideration though.

**Comment by omarsy** [comment]
```
My initial inclination would be to add a DEV-only check + warning that symbols-as-keys isn't supported.

That being said, I don't really speak for the team on this particular issue 😄
```

I think we have to deal with symbols as we deal with objects. At the moment when we add an object as a key we don't get an error message or a warning we get the key which becomes `[object Object]`.

**Comment by omarsy** [comment]
But I think that if javascript supports Symbol as the key to an object we should also support it. But maybe I'm wrong

**Comment by bpernick** [comment]
Any further thoughts on this?

**Comment by johnloven** [comment]
> At the moment when we add an object as a key we don't get an error message or a warning we get the key which becomes `[object Object]`.

As a React developer for the last couple of years I did not know this was the case. And the [docs](https://reactjs.org/docs/lists-and-keys.html#keys) doesn't really make that clear either. I think it would be helpful to show a dev warning for all non-primitive keys, as well as make it clear in the docs that all keys are converted to strings.

**Comment by bpernick** [comment]
Yeah I like showing a dev warning for any key that isn't a string or a number. But I'd still be in favor of stringifying Symbols as well as everything else just for consistency's sake

**Comment by bpernick** [comment]
So in addition to the things below, I added a dev warning for any type of key other than a string or number. What do yall think of that approach?

**Comment by bvaughn** [comment]
Someone will have to review the PR, @bpernick. This often takes a while as there are only a couple of core contributors and we're juggling a lot of things. 🙇  Sorry.
